forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/ApplicationPinDeviceAuthenticator.kt

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2022 ForgeRock. All rights reserved.
+ *
+ * This software may be modified and distributed under the terms
+ * of the MIT license. See the LICENSE file for details.
+ */
+package org.forgerock.android.auth.devicebind
+
+import android.content.Context
+import android.os.OperationCanceledException
+import androidx.annotation.VisibleForTesting
+import androidx.fragment.app.FragmentActivity
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlinx.coroutines.yield
+import org.forgerock.android.auth.AppPinAuthenticator
+import org.forgerock.android.auth.CryptoKey
+import org.forgerock.android.auth.EncryptedFileKeyStore
+import org.forgerock.android.auth.InitProvider
+import org.forgerock.android.auth.KeyStoreRepository
+import org.forgerock.android.auth.Logger
+import org.forgerock.android.auth.callback.DeviceBindingAuthenticationType
+import java.io.FileNotFoundException
+import java.io.InputStream
+import java.io.OutputStream
+import java.security.KeyStore
+import java.security.PrivateKey
+import java.security.UnrecoverableKeyException
+import java.security.interfaces.RSAPublicKey
+import java.util.*
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * Device Authenticator which use Application PIN to secure device cryptography keys
+ */
+open class ApplicationPinDeviceAuthenticator : CryptoAware, DeviceAuthenticator,
+    KeyStoreRepository {
+
+    @VisibleForTesting
+    internal lateinit var appPinAuthenticator: AppPinAuthenticator
+    internal lateinit var keyStore: KeyStoreRepository
+
+    private val pinSuffix = "_PIN"
+
+    protected lateinit var prompt: Prompt
+
+    @VisibleForTesting
+    internal val pinRef = AtomicReference<CharArray>()
+
+    internal val worker = Executors.newSingleThreadScheduledExecutor()
+
+    override suspend fun generateKeys(context: Context): KeyPair {
+        val pin = requestForCredentials()
+        yield() //Suspension point for timeout, we don't want to create the key if timeout
```

Comment:
is there a way to avoid this yield ? something like async await 

Example:
withContext(Dispatchers.Default) { requestForCredentials() }

or 

 async { requestForCredentials() }.await()

Reply:
Yes, the yield is not required for suspension point with continuation. The continuation can handle the timeout.